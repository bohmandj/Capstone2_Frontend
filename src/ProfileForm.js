import React, { useState, useContext } from 'react';
import { Navigate, useNavigate } from 'react-router-dom';
import MemoLedgerContext from './MemoLedgerContext';
import MemoLedgerApi from './api';
import {
    Button,
    CardTitle,
    Form,
    FormFeedback,
    FormGroup,
    Input,
    Label
} from "reactstrap";

const ProfileForm = ({ setShowProfileForm }) => {
    /* Profile form to update user data */

    const { currentUser, setCurrentUser, setIsLoading } = useContext(MemoLedgerContext);
    const navigate = useNavigate();

    const ERRORS_INITIAL_STATE = {
        invalidEmail: false,
        shortPassword: false,
        unmatchedPassword: false,
    }
    const [errors, setErrors] = useState(ERRORS_INITIAL_STATE);
    const [touched, setTouched] = useState({
        email: false,
        password: false,
        confirmPassword: false,
    });

    const inputs = [
        "email",
        "password",
        "confirmPassword"
    ]
    const INPUTS_INITIAL_STATE = inputs.reduce((obj, input) => {
        obj[input] = currentUser[input] || "";
        return obj;
    }, {});

    const [formData, setFormData] = useState(INPUTS_INITIAL_STATE);

    if (!currentUser) {
        return <Navigate to={'/'} />
    }

    const updateUser = async (profileFormData) => {
        setIsLoading(true);
        const {
            password,
            email
        } = profileFormData;
        const returnedUser = await MemoLedgerApi.updateUser(currentUser.username, password, email);
        setCurrentUser({
            ...currentUser,
            email: returnedUser.email
        });
        setIsLoading(false);
        navigate('/');
    }

    const validateEmail = (email) => {
        return email.includes('@');
    };
    const validatePasswordMatch = (pwd, confirmPwd) => {
        return pwd === confirmPwd;
    };
    const validatePasswordLength = (pwd) => {
        return pwd.length >= 8;
    }

    const validateField = (name, value, formValues = formData) => {
        switch (name) {
            case 'email':
                return { invalidEmail: !validateEmail(value) };
            case 'password':
                return {
                    shortPassword: !validatePasswordLength(value),
                    unmatchedPassword: !validatePasswordMatch(value, formValues.confirmPassword),
                };
            case 'confirmPassword':
                return {
                    unmatchedPassword: !validatePasswordMatch(formValues.password, value),
                };
            default:
                return {};
        }
    };

    const handleChange = (e) => {
        const { name, value } = e.target;
        const updatedFormData = {
            ...formData,
            [name]: value
        };
        setFormData(updatedFormData);

        if (touched[name]) {
            const fieldErrors = validateField(name, value, updatedFormData);
            setErrors(prev => ({ ...prev, ...fieldErrors }));
        }
    };

    const handleBlur = (e) => {
        const { name } = e.target;
        setTouched((t) => ({ ...t, [name]: true }));

        const fieldErrors = validateField(name, formData[name]);
        setErrors(prev => ({ ...prev, ...fieldErrors }));
    };

    const handleSubmit = async (e) => {
        e.preventDefault();

        const validateAll = () => {
            return {
                invalidEmail: !validateEmail(formData.email),
                shortPassword: !validatePasswordLength(formData.password),
                unmatchedPassword: !validatePasswordMatch(formData.password, formData.confirmPassword),
            };
        };

        const newErrors = validateAll();
        setErrors(newErrors);
        setTouched({
            email: true,
            password: true,
            confirmPassword: true,
        });

        const isValid = Object.values(newErrors).every(error => error === false);

        if (isValid) {
            updateUser(formData);
        }
    };

    const handleCancel = () => {
        setFormData(INPUTS_INITIAL_STATE);
        setErrors(ERRORS_INITIAL_STATE);
        setShowProfileForm(false);
    }

    return (<>
        <CardTitle tag="h3">
            Edit Profile
        </CardTitle>
        <Form onSubmit={handleSubmit}>
            <FormGroup>
                <Label
                    className='font-weight-bold'
                    htmlFor='username'
                >Username:</Label>
                <Input
                    id="username"
                    type="text"
                    name="username"
                    placeholder={currentUser.username}
                    disabled
                    readOnly
                />
            </FormGroup>

            <FormGroup>
                <Label
                    className='font-weight-bold'
                    htmlFor='email'
                >Email:</Label>
                <Input
                    type="email"
                    name="email"
                    value={formData.email}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    invalid={errors.invalidEmail && touched.email}
                />
                {errors.invalidEmail && touched.email && (
                    <FormFeedback>Invalid Email</FormFeedback>
                )}
            </FormGroup>

            <FormGroup>
                <Label
                    className='font-weight-bold'
                    htmlFor='password'
                >New Password:</Label>
                <Input
                    type="password"
                    name="password"
                    value={formData.password}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    invalid={errors.shortPassword && touched.password}
                />
                {errors.shortPassword && touched.password
                    && <FormFeedback>
                        Password must be at least 8 characters
                    </FormFeedback>
                }
            </FormGroup>

            <FormGroup>
                <Label
                    className='font-weight-bold'
                    htmlFor='confirmPassword'
                >Confirm New Password:</Label>
                <Input
                    type="password"
                    name="confirmPassword"
                    value={formData.confirmPassword}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    invalid={errors.unmatchedPassword
                        && (touched.password || touched.confirmPassword)
                    }
                />
                {errors.unmatchedPassword
                    && (touched.password || touched.confirmPassword)
                    && <FormFeedback>
                        Passwords must match
                    </FormFeedback>
                }
            </FormGroup>

            <Button name="submit" className='w-100' color='success'>Submit</Button>
            <Button className='w-100 mt-2' onClick={() => handleCancel()}>Cancel</Button>
        </Form>
    </>
    )
}

export default ProfileForm;