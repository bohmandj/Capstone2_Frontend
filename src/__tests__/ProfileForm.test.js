import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import ProfileForm from '../ProfileForm';
import MemoLedgerContext from '../MemoLedgerContext';
import MemoLedgerApi from '../api';

// Mock API calls
jest.mock('../api');

// Mock context values
const mockCurrentUser = {
    username: 'testuser',
    email: 'test@user.com'
};

const mockSetCurrentUser = jest.fn();
const mockSetIsLoading = jest.fn();
const mockSetShowProfileForm = jest.fn();

const mockContextValueLoggedIn = {
    currentUser: mockCurrentUser,
    setCurrentUser: mockSetCurrentUser,
    setIsLoading: mockSetIsLoading,
};

const mockContextValueLoggedOut = {
    currentUser: null,
    setCurrentUser: mockSetCurrentUser,
    setIsLoading: mockSetIsLoading,
};

// Helper function to render ProfileForm within context and router
const renderProfileForm = (contextValue, props = {}, initialEntries = ['/profile']) => {
    return render(
        <MemoLedgerContext.Provider value={contextValue}>
            <MemoryRouter initialEntries={initialEntries}>
                <ProfileForm {...props} />
            </MemoryRouter>
        </MemoLedgerContext.Provider>
    );
};

describe('ProfileForm', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('renders without crashing when logged in', () => {
        renderProfileForm(mockContextValueLoggedIn, { setShowProfileForm: mockSetShowProfileForm });
        expect(screen.getByText('Edit Profile')).toBeInTheDocument();
    });

    test('displays correct initial elements when logged in', () => {
        renderProfileForm(mockContextValueLoggedIn, { setShowProfileForm: mockSetShowProfileForm });

        // Username should be disabled and show current username in placeholder
        const usernameInput = screen.getByLabelText(/username:/i);
        expect(usernameInput).toBeInTheDocument();
        expect(usernameInput).toHaveAttribute('placeholder', mockCurrentUser.username);
        expect(usernameInput).toBeDisabled();

        // Email should show current email
        const emailInput = screen.getByLabelText(/email:/i);
        expect(emailInput).toBeInTheDocument();
        expect(emailInput).toHaveValue(mockCurrentUser.email);

        // Password fields should be empty and of type password
        const newPasswordInput = screen.getByLabelText('New Password:');
        const confirmPasswordInput = screen.getByLabelText('Confirm New Password:');

        expect(newPasswordInput).toHaveAttribute('type', 'password');
        expect(newPasswordInput).toHaveValue('');
        expect(confirmPasswordInput).toHaveAttribute('type', 'password');
        expect(confirmPasswordInput).toHaveValue('');

        // Buttons
        expect(screen.getByRole('button', { name: /submit/i })).toBeInTheDocument();
        expect(screen.getByRole('button', { name: /cancel/i })).toBeInTheDocument();
    });

    test('calls updateUser and updates context on successful submission', async () => {
        MemoLedgerApi.updateUser.mockResolvedValue({ username: 'testuser', email: 'new@email.com' });

        renderProfileForm(mockContextValueLoggedIn, { setShowProfileForm: mockSetShowProfileForm });

        const emailInput = screen.getByDisplayValue(mockCurrentUser.email);
        const passwordInput = screen.getByLabelText('New Password:');
        const confirmPasswordInput = screen.getByLabelText('Confirm New Password:');
        const submitButton = screen.getByRole('button', { name: /submit/i });

        fireEvent.change(emailInput, { target: { value: 'new@email.com' } });
        fireEvent.change(passwordInput, { target: { value: 'newpassword' } });
        fireEvent.change(confirmPasswordInput, { target: { value: 'newpassword' } });
        fireEvent.click(submitButton);

        await waitFor(() => {
            expect(mockSetIsLoading).toHaveBeenCalledWith(true);
            expect(MemoLedgerApi.updateUser).toHaveBeenCalledWith(
                mockCurrentUser.username,
                'newpassword',
                'new@email.com'
            );
            expect(mockSetCurrentUser).toHaveBeenCalledWith({
                ...mockCurrentUser,
                email: 'new@email.com'
            });
            expect(mockSetIsLoading).toHaveBeenCalledWith(false);
        });
    });

    test('shows error and prevents submission for invalid email', async () => {
        renderProfileForm(mockContextValueLoggedIn, { setShowProfileForm: mockSetShowProfileForm });

        const emailInput = screen.getByLabelText(/email:/i);
        const submitButton = screen.getByRole('button', { name: /submit/i });

        fireEvent.change(emailInput, { target: { value: 'invalid-email' } });
        fireEvent.blur(emailInput);

        expect(await screen.findByText('Invalid Email')).toBeInTheDocument();

        fireEvent.click(submitButton);

        expect(MemoLedgerApi.updateUser).not.toHaveBeenCalled();
    });

    test('shows error and prevents submission for short password', async () => {
        renderProfileForm(mockContextValueLoggedIn, { setShowProfileForm: mockSetShowProfileForm });

        const passwordInput = screen.getByLabelText('New Password:');
        const confirmPasswordInput = screen.getByLabelText('Confirm New Password:');
        const submitButton = screen.getByRole('button', { name: /submit/i });

        fireEvent.change(passwordInput, { target: { value: 'short' } });
        fireEvent.blur(passwordInput);
        fireEvent.change(confirmPasswordInput, { target: { value: 'short' } });
        fireEvent.blur(confirmPasswordInput);

        expect(await screen.findByText('Password must be at least 8 characters')).toBeInTheDocument();

        fireEvent.click(submitButton);

        expect(MemoLedgerApi.updateUser).not.toHaveBeenCalled();
    });

    test('shows error and prevents submission for mismatched passwords', async () => {
        renderProfileForm(mockContextValueLoggedIn, { setShowProfileForm: mockSetShowProfileForm });

        const passwordInput = screen.getByLabelText('New Password:');
        const confirmPasswordInput = screen.getByLabelText('Confirm New Password:');
        const submitButton = screen.getByRole('button', { name: /submit/i });

        fireEvent.change(passwordInput, { target: { value: 'longenough' } });
        fireEvent.blur(passwordInput);
        fireEvent.change(confirmPasswordInput, { target: { value: 'nomatch' } });
        fireEvent.blur(confirmPasswordInput);

        expect(await screen.findByText('Passwords must match')).toBeInTheDocument();

        fireEvent.click(submitButton);

        expect(MemoLedgerApi.updateUser).not.toHaveBeenCalled();
    });

    test('resets form and calls setShowProfileForm(false) on cancel', async () => {
        renderProfileForm(mockContextValueLoggedIn, { setShowProfileForm: mockSetShowProfileForm });

        const emailInput = screen.getByDisplayValue(mockCurrentUser.email);
        const passwordInput = screen.getByLabelText('New Password:');
        const cancelButton = screen.getByRole('button', { name: /cancel/i });

        // Change some data
        fireEvent.change(emailInput, { target: { value: 'changed@email.com' } });
        fireEvent.change(passwordInput, { target: { value: 'somenewpassword' } });

        // Click cancel
        fireEvent.click(cancelButton);

        // Expect form data to revert to initial state
        await waitFor(() => {
            expect(emailInput).toHaveValue(mockCurrentUser.email);
            expect(passwordInput).toHaveValue('');
            expect(mockSetShowProfileForm).toHaveBeenCalledWith(false);
        });

        // Ensure updateUser was NOT called
        expect(MemoLedgerApi.updateUser).not.toHaveBeenCalled();
    });

    test('matches snapshot when logged in', () => {
        const { asFragment } = renderProfileForm(mockContextValueLoggedIn, { setShowProfileForm: mockSetShowProfileForm });
        expect(asFragment()).toMatchSnapshot();
    });
});