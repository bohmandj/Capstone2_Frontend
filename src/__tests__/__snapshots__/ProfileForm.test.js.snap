// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileForm matches snapshot when logged in 1`] = `
<DocumentFragment>
  <h3
    class="card-title"
  >
    Edit Profile
  </h3>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="font-weight-bold form-label"
        for="username"
      >
        Username:
      </label>
      <input
        class="form-control"
        disabled=""
        id="username"
        name="username"
        placeholder="testuser"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="font-weight-bold form-label"
        for="email"
      >
        Email:
      </label>
      <input
        aria-invalid="false"
        class="form-control"
        id="email"
        name="email"
        type="email"
        value="test@user.com"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="font-weight-bold form-label"
        for="password"
      >
        New Password:
      </label>
      <input
        aria-invalid="false"
        class="form-control"
        id="password"
        name="password"
        type="password"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="font-weight-bold form-label"
        for="confirmPassword"
      >
        Confirm New Password:
      </label>
      <input
        aria-invalid="false"
        class="form-control"
        id="confirmPassword"
        name="confirmPassword"
        type="password"
        value=""
      />
    </div>
    <button
      class="w-100 btn btn-success"
      name="submit"
    >
      Submit
    </button>
    <button
      class="w-100 mt-2 btn btn-secondary"
      type="button"
    >
      Cancel
    </button>
  </form>
</DocumentFragment>
`;
